pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,305 @@
+			return false, err
+		}
+
+		if shouldWrite {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// shouldWriteFile returns whether a new version of a file should be written to disk.
+func (w *AtomicWriter) shouldWriteFile(path string, content []byte) (bool, error) {
+	_, err := os.Lstat(path)
+	if err != nil && os.IsNotExist(err) {
+		return true, nil
+	}
+
+	contentOnFs, err := ioutil.ReadFile(path)
+	if err != nil {
+		return false, err
+	}
+
+	return (bytes.Compare(content, contentOnFs) != 0), nil
+}
+
+// newTimestampDir creates a new timestamped directory
+func (w *AtomicWriter) newTimestampDir() (string, error) {
+	tsDir, err := ioutil.TempDir(w.TargetDir, fmt.Sprintf("..%v", ts()))
+	if err != nil {
+		glog.Errorf("%v: unable to create new temp directory: %v", w.LogContext, err)
+		return "", err
+	}
+
+	return tsDir, nil
+}
+
+// writePayloadToDir writes the given payload to the given directory.  The
+// directory must exist.
+func (w *AtomicWriter) writePayloadToDir(payload map[string][]byte, dir string) error {
+	allErrs := []error{}
+	for projectionPath, content := range payload {
+		fullPath := path.Join(dir, path.Clean(projectionPath))
+		dir, _ := filepath.Split(fullPath)
+
+		err := os.MkdirAll(dir, os.ModePerm)
+		if err != nil {
+			glog.Errorf("%v: unable to create directory %v: %v", w.LogContext, dir, err)
+			allErrs = append(allErrs, err)
+		}
+
+		err = ioutil.WriteFile(fullPath, content, 0644)
+		if err != nil {
+			glog.Errorf("%v: unable to write file %v: %v", w.LogContext, fullPath, err)
+			allErrs = append(allErrs, err)
+		}
+	}
+
+	return utilerrors.NewAggregate(allErrs)
+}
+
+// createPayloadSymlinksInTargetDir creates the relative symlinks for all the
+// files configured in the payload. If the directory in a file path does not
+// exist, it is created.
+//
+// Viz:
+// For files: "bar", "foo/bar", "baz/bar", "foo/baz/blah"
+// the following symlinks and subdirectories are created:
+// bar          -> ..identifierToken/bar
+// foo/bar      -> ../..identifierToken/foo/bar
+// baz/bar      -> ../..identifierToken/baz/bar
+// foo/baz/blah -> ../../..identifierToken/foo/baz/blah
+func (w *AtomicWriter) createPayloadSymlinksInTargetDir(payload map[string][]byte) error {
+	for projectionPath := range payload {
+		cleanProjectionPath := path.Clean(projectionPath)
+		dir, _ := filepath.Split(cleanProjectionPath)
+		subDirs := 0
+		if len(dir) > 0 {
+			// If dir is not empty, the projection path contains at least one
+			// subdirectory (example: projectionPath := "foo/bar").
+			// Since filepath.Split leaves a trailing path separator, in this
+			// example, dir = "foo/".  In order to calculate the number of
+			// subdirectories, we must subtract 1 from the number returned by split.
+			subDirs = len(strings.Split(dir, "/")) - 1
+			err := os.MkdirAll(path.Join(w.TargetDir, dir), os.ModePerm)
+			if err != nil {
+				return err
+			}
+		}
+		_, err := os.Readlink(path.Join(w.TargetDir, cleanProjectionPath))
```

Comment:
@sdminonne I intentionally avoided that form of if statement because my own subjective opinion is that it makes the whole thing harder to read and understand.  Excuse my feeble brain :)